File: src/efficio_pptx_contracts/classification_schemes.py
"""Typed presentation-level classification schemes for categorical fills."""

from __future__ import annotations

import json
from collections.abc import Mapping
from dataclasses import dataclass
from enum import StrEnum
from typing import Any, cast

from jsonschema import Draft202012Validator

from ._resources import load_json

CLASSIFICATION_SCHEMES_TAG = "efficio_classification_schemes"
CLASSIFICATION_SCHEME_ID_TAG = "efficio_classification_scheme_id"


class ClassificationPaletteMode(StrEnum):
    """Supported trusted palette representations."""

    RGB = "rgb"


@dataclass(frozen=True)
class ClassificationFill:
    """One authored solid fill in the exact deck-tag wire shape."""

    value: str


@dataclass(frozen=True)
class ClassificationCase:
    """One semantic case and its trusted presentation styling."""

    case_id: str
    label: str
    description: str
    fill: ClassificationFill


@dataclass(frozen=True)
class ClassificationScheme:
    """One ordered reusable semantic classification scheme."""

    scheme_id: str
    instruction: str
    palette_mode: ClassificationPaletteMode
    cases: tuple[ClassificationCase, ...]


def parse_classification_schemes(
    raw_value: str | list[Any] | None,
) -> tuple[ClassificationScheme, ...]:
    """Parse and semantically validate a deck classification-schemes value."""
    if raw_value is None or (isinstance(raw_value, str) and not raw_value.strip()):
        return ()
    parsed = _decode_value(raw_value)
    errors = sorted(
        Draft202012Validator(_load_scheme_schema()).iter_errors(parsed),
        key=lambda error: tuple(str(part) for part in error.absolute_path),
    )
    if errors:
        raise ValueError("classification schemes do not match the deck tag contract")

    records = cast(list[dict[str, Any]], parsed)
    schemes = tuple(_parse_scheme(record) for record in records)
    _validate_unique_ids(schemes)
    return schemes


def resolve_categorical_fill_scheme(
    component_tags: Mapping[str, str],
    deck_tags: Mapping[str, str],
) -> ClassificationScheme:
    """Resolve one categorical-fill component's referenced authored scheme."""
    raw_scheme_id = component_tags.get(CLASSIFICATION_SCHEME_ID_TAG)
    if not isinstance(raw_scheme_id, str) or not raw_scheme_id.strip():
        raise ValueError(
            f"categorical-fill component requires {CLASSIFICATION_SCHEME_ID_TAG}"
        )
    scheme_id = raw_scheme_id.strip()
    schemes = parse_classification_schemes(deck_tags.get(CLASSIFICATION_SCHEMES_TAG))
    for scheme in schemes:
        if scheme.scheme_id == scheme_id:
            return scheme
    raise ValueError("categorical-fill component references an unknown classification scheme")


def classification_scheme_public_context(
    scheme: ClassificationScheme,
) -> dict[str, Any]:
    """Return semantic AI context without trusted fill values."""
    return {
        "scheme_id": scheme.scheme_id,
        "instruction": scheme.instruction,
        "cases": [
            {
                "case_id": case.case_id,
                "label": case.label,
                "description": case.description,
            }
            for case in scheme.cases
        ],
    }


def classification_scheme_private_metadata(
    scheme: ClassificationScheme,
) -> dict[str, Any]:
    """Return deterministic trusted metadata for validation and rendering."""
    return {
        "scheme_id": scheme.scheme_id,
        "cases": {
            case.case_id: {
                "fill": {"kind": scheme.palette_mode.value, "value": case.fill.value}
            }
            for case in scheme.cases
        },
    }


def _decode_value(raw_value: str | list[Any]) -> Any:
    if not isinstance(raw_value, str):
        return raw_value
    try:
        return json.loads(raw_value)
    except json.JSONDecodeError as error:
        raise ValueError("classification schemes must contain valid JSON") from error


def _load_scheme_schema() -> dict[str, Any]:
    contract = load_json("schemas", "presentation", "deck-tags.json")
    definition = contract.get("tags", {}).get(CLASSIFICATION_SCHEMES_TAG)
    if not isinstance(definition, dict) or not isinstance(
        definition.get("schema"), dict
    ):
        raise ValueError("classification-schemes deck tag contract is unavailable")
    return cast(dict[str, Any], definition["schema"])


def _parse_scheme(record: dict[str, Any]) -> ClassificationScheme:
    mode = ClassificationPaletteMode(record["palette_mode"])
    return ClassificationScheme(
        scheme_id=record["scheme_id"],
        instruction=record["instruction"].strip(),
        palette_mode=mode,
        cases=tuple(
            ClassificationCase(
                case_id=case["case_id"],
                label=case["label"].strip(),
                description=case["description"].strip(),
                fill=ClassificationFill(value=case["fill"]["value"]),
            )
            for case in record["cases"]
        ),
    )


def _validate_unique_ids(schemes: tuple[ClassificationScheme, ...]) -> None:
    scheme_ids: set[str] = set()
    for scheme in schemes:
        if scheme.scheme_id in scheme_ids:
            raise ValueError("classification scheme IDs must be unique")
        scheme_ids.add(scheme.scheme_id)
        case_ids: set[str] = set()
        for case in scheme.cases:
            if case.case_id in case_ids:
                raise ValueError(
                    f"case IDs must be unique within scheme {scheme.scheme_id!r}"
                )
            case_ids.add(case.case_id)


__all__ = [
    "CLASSIFICATION_SCHEMES_TAG",
    "CLASSIFICATION_SCHEME_ID_TAG",
    "ClassificationCase",
    "ClassificationFill",
    "ClassificationPaletteMode",
    "ClassificationScheme",
    "classification_scheme_private_metadata",
    "classification_scheme_public_context",
    "parse_classification_schemes",
    "resolve_categorical_fill_scheme",
]

File: ts/editor/classification-schemes.ts
import { deckTagSchema } from "../../generated/ts/presentation/deckTagSchema.js";

export const CLASSIFICATION_SCHEMES_TAG = "efficio_classification_schemes";
export const CLASSIFICATION_SCHEME_ID_TAG = "efficio_classification_scheme_id";

export type ClassificationPaletteMode = "rgb";

export type ClassificationFill = {
  value: string;
};

export type ClassificationCase = {
  case_id: string;
  label: string;
  description: string;
  fill: ClassificationFill;
};

export type ClassificationScheme = {
  scheme_id: string;
  instruction: string;
  palette_mode: ClassificationPaletteMode;
  cases: readonly ClassificationCase[];
};

export class ClassificationSchemeContractError extends Error {
  override readonly name = "ClassificationSchemeContractError";
}

const definition = deckTagSchema.tags.efficio_classification_schemes.schema;
const schemeDefinition = definition.items;
const caseDefinition = schemeDefinition.properties.cases;
const idPattern = new RegExp(schemeDefinition.properties.scheme_id.pattern);
const rgbPattern = new RegExp(
  caseDefinition.items.properties.fill.properties.value.pattern
);

export function parseClassificationSchemes(
  rawValue: string | readonly unknown[] | undefined
): readonly ClassificationScheme[] {
  if (rawValue === undefined || (typeof rawValue === "string" && rawValue.trim() === "")) {
    return [];
  }
  const parsed = decodeValue(rawValue);
  if (!Array.isArray(parsed) || parsed.length > definition.maxItems) {
    throw invalidSchemes();
  }

  const schemes = parsed.map(parseScheme);
  requireUnique(schemes.map((scheme) => scheme.scheme_id), "scheme IDs");
  return schemes;
}

export function resolveCategoricalFillScheme(
  componentTags: Readonly<Record<string, string>>,
  deckTags: Readonly<Record<string, string>>
): ClassificationScheme {
  const schemeId = componentTags[CLASSIFICATION_SCHEME_ID_TAG]?.trim();
  if (!schemeId) {
    throw new ClassificationSchemeContractError(
      `Categorical-fill components require ${CLASSIFICATION_SCHEME_ID_TAG}.`
    );
  }
  const scheme = parseClassificationSchemes(deckTags[CLASSIFICATION_SCHEMES_TAG]).find(
    (candidate) => candidate.scheme_id === schemeId
  );
  if (!scheme) {
    throw new ClassificationSchemeContractError(
      "The categorical-fill component references an unknown classification scheme."
    );
  }
  return scheme;
}

function decodeValue(rawValue: string | readonly unknown[]): unknown {
  if (typeof rawValue !== "string") return rawValue;
  try {
    return JSON.parse(rawValue) as unknown;
  } catch {
    throw new ClassificationSchemeContractError(
      "Classification schemes must contain valid JSON."
    );
  }
}

function parseScheme(value: unknown): ClassificationScheme {
  if (!isExactObject(value, ["scheme_id", "instruction", "palette_mode", "cases"])) {
    throw invalidSchemes();
  }
  const schemeId = requiredText(value.scheme_id, 120, idPattern);
  const instruction = requiredText(value.instruction, 2000);
  if (value.palette_mode !== "rgb" || !Array.isArray(value.cases)) {
    throw invalidSchemes();
  }
  if (
    value.cases.length < caseDefinition.minItems ||
    value.cases.length > caseDefinition.maxItems
  ) {
    throw invalidSchemes();
  }
  const cases = value.cases.map(parseCase);
  requireUnique(cases.map((item) => item.case_id), `case IDs in ${schemeId}`);
  return {
    scheme_id: schemeId,
    instruction,
    palette_mode: "rgb",
    cases,
  };
}

function parseCase(value: unknown): ClassificationCase {
  if (!isExactObject(value, ["case_id", "label", "description", "fill"])) {
    throw invalidSchemes();
  }
  if (!isExactObject(value.fill, ["value"])) {
    throw invalidSchemes();
  }
  const fillValue = requiredText(value.fill.value, 6, rgbPattern);
  return {
    case_id: requiredText(value.case_id, 120, idPattern),
    label: requiredText(value.label, 120),
    description: requiredText(value.description, 1000),
    fill: { value: fillValue },
  };
}

function requiredText(value: unknown, maximum: number, pattern?: RegExp): string {
  if (
    typeof value !== "string" ||
    value.trim() === "" ||
    [...value].length > maximum ||
    (pattern !== undefined && !pattern.test(value))
  ) {
    throw invalidSchemes();
  }
  return value.trim();
}

function isExactObject(
  value: unknown,
  keys: readonly string[]
): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  return actual.length === expected.length && actual.every((key, index) => key === expected[index]);
}

function requireUnique(values: readonly string[], subject: string): void {
  if (new Set(values).size !== values.length) {
    throw new ClassificationSchemeContractError(`Classification ${subject} must be unique.`);
  }
}

function invalidSchemes(): ClassificationSchemeContractError {
  return new ClassificationSchemeContractError(
    "Classification schemes do not match the deck tag contract."
  );
}
